### PR TITLE
Teams redesign

### PR DESCRIPTION
**Pull request by iskhakov**
# What this PR does

* api returns all the resources available to the user by default 
* substitutes `team switcher` with `multi-select team filter`
* allow referencing between integrations - escalations chains - [schedules, outgoing webhooks] across teams

https://user-images.githubusercontent.com/2262529/225634581-2d2e8af2-15ce-4c01-a90e-8267d98f5a23.mov

## Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated

### CODE REVIEW

File: engine/apps/alerts/models/alert_receive_channel.py [hunk 1]
```diff
@@ -235,55 +235,16 @@ def hard_delete(self):
...
             raise TeamCanNotBeChangedError("Integration is already in this team")
 
         if team_id is not None:
-            new_team = user.teams.filter(public_primary_key=team_id).first()
+            new_team = user.available_teams().filter(public_primary_key=team_id).first()
             if not new_team:
                 raise TeamCanNotBeChangedError("User is not a member of the selected team")
         else:
             new_team = None  # means General team
-
```
Comment (iskhakov) [inline engine/apps/alerts/models/alert_receive_channel.py:249]: We don't need this check anymore because we allow connect resources from different teams

File: engine/apps/api/views/alert_group.py [hunk 2]
```diff
@@ -90,6 +95,14 @@ class AlertGroupFilter(DateRangeFilterMixin, ModelFieldFilterMixin, filters.Filt
     )
     with_resolution_note = filters.BooleanFilter(method="filter_with_resolution_note")
     mine = filters.BooleanFilter(method="filter_mine")
+    team = filters.ModelMultipleChoiceFilter(
```
Comment (iskhakov) [inline engine/apps/api/views/alert_group.py:98]: Here and below I use the same approach we have for public api, but modified to support multi value filters

File: engine/apps/api/serializers/team.py [hunk 3]
```diff
@@ -13,4 +13,12 @@ class Meta:
             "name",
             "email",
             "avatar_url",
+            "is_sharing_resources_to_all",
```
Comment (iskhakov) [inline engine/apps/api/serializers/team.py:16]: we need to be able to modify only `is_sharing_resources_to_all`

File: engine/apps/user_management/models/user.py [hunk 4]
```diff
@@ -183,6 +183,12 @@ class Meta:
     def __str__(self):
         return f"{self.pk}: {self.username}"
 
+    @property
+    def available_teams(self):
```
Comment (iskhakov) [inline engine/apps/user_management/models/user.py:187]: property that returns teams where user has access to based on membership, team is shared
if user is admin they are able to see all the org teams

File: engine/common/api_helpers/filters.py [hunk 5]
```diff
@@ -69,6 +70,22 @@ def filter_model_field_with_single_value(self, queryset, name, value):
         queryset = queryset.filter(**lookup_kwargs)
         return queryset
 
+    def filter_model_field_with_multiple_values(self, queryset, name, values):
```
Comment (iskhakov) [inline engine/common/api_helpers/filters.py:73]: adding a method to filter on multiple teams
we reuse the same mixin we use for public api

**Review by vadimkerr — APPROVED** [review]
This is super impressive, I LOVE to see that the team switcher is finally removed 🎉 
Backend changes look good to me, only a few minor comments.

Also some questions:
1. When loading a page with the team filter pre-selected, it first loads all resources from org and only then re-renders the page with relevant results, which makes all the pages look a bit glitchy. Is this expected?
2. Does the user page show all users from org? I'm not sure which users are shown and there's no info on that. Probably treating users as all other resources would make sense? For example, it could be useful to filter users by team and see their notification policies IMO.

**Review by vadimkerr — CHANGES_REQUESTED** [review]
I think I found a bug, can't create resources with "no team" selected:

Screenshot 2023-03-16 at 18 15 32

Screenshot 2023-03-16 at 18 15 55

Requesting changes to address this.

File: grafana-plugin/src/containers/TeamName/TeamName.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      ) : (
+        <Text type="secondary">
+          <Avatar size="small" src={team.avatar_url} />{' '}
+          <Tooltip placement="top" content={'Resource is assigned to ' + team.name}>
+            <span>{team.name}</span>
+          </Tooltip>{' '}
+        </Text>
+      )}
+    </>
+  );
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/TeamName/TeamName.tsx:29]:
```suggestion
  if (team.id === 'null') {
    return <Badge text={team.name} color={'blue'} tooltip={'Resource is not assigned to any team (ex General team)'} />;
  }
  return (
    <Text type="secondary">
      <Avatar size="small" src={team.avatar_url} />{' '}
      <Tooltip placement="top" content={'Resource is assigned to ' + team.name}>
        <span>{team.name}</span>
      </Tooltip>{' '}
    </Text>
  );
```

Few other questions:
- should `team.id` be the literal string of `'null'` or actual `null`?
- we are getting rid of the idea of the "General" team, right? If so, consider rewording the tooltip message.

File: grafana-plugin/src/containers/TeamsList/TeamsList.tsx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+      (record.id === 'null' && store.userStore.currentUser?.current_team === null) ||
+      record.id === store.userStore.currentUser?.current_team
+    );
+  };
+
+  const renderTeam = (record: GrafanaTeam) => {
+    return (
+      <>
+        {record.id === 'null' ? (
+          <Badge text={record.name} color={'blue'} tooltip={'Resource is not assigned to any team (ex General team)'} />
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/TeamsList/TeamsList.tsx:30]: similar comment here regarding literal `'null'` string and "General" team name

File: grafana-plugin/src/containers/TeamsList/TeamsList.tsx [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            <Avatar size="small" src={record.avatar_url} /> {record.name}{' '}
+          </Text>
+        )}
+        {isTeamDefault(record) && (
+          <>
+            {' '}
+            <Tooltip
+              content={
+                (record.id === 'null' ? 'No team' : 'This team') +
+                ` will be selected by default when creating new resources (Integrations, Escalation Chains, Schedules, Outgoing Webhooks)`
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/TeamsList/TeamsList.tsx:42]: _nit_
```suggestion
                ` will be selected by default when creating new resources`
```

File: grafana-plugin/src/containers/TeamsList/TeamsList.tsx [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+      width: '15%',
+      key: 'action',
+      render: renderActionButtons,
+    },
+  ];
+
+  return (
+    <>
+      <GTable
+        // emptyText={initialUsersLoaded ? 'No users found' : 'Loading...'}
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/TeamsList/TeamsList.tsx:118]: can this line be removed?

File: grafana-plugin/src/models/alert_receive_channel/alert_receive_channel.ts [hunk 10]
```diff
@@ -81,8 +81,12 @@ export class AlertReceiveChannelStore extends BaseStore {
...
 
   @action
-  async updateItems(query = '') {
-    const result = await this.getAll(query);
+  async updateItems(query: any = '') {
+    const params = typeof query === 'string' ? { search: query } : query;
+
+    const result = await makeRequest(`${this.path}`, {
+      params,
+    });
```
Comment (joeyorlando) [inline grafana-plugin/src/models/alert_receive_channel/alert_receive_channel.ts:89]:
```suggestion
    const result = await makeRequest(this.path, { params });
```

File: grafana-plugin/src/models/filters/filters.types.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import { SelectOption } from 'state/types';
+
+export interface FiltersValues {
+  [key: string]: any;
+}
+
+export interface FilterOption {
+  name: string;
+  type: 'search' | 'options' | 'boolean' | 'daterange' | 'team_select';
```
Comment (joeyorlando) [inline grafana-plugin/src/models/filters/filters.types.ts:9]: I think I saw `'search' | 'options' | 'boolean' | 'daterange' | 'team_select'` defined somewhere else as well, can this be extracted higher up to something like this, and reused?
```ts
export type FilterOptionType = 'search' | 'options' | 'boolean' | 'daterange' | 'team_select';
```

File: grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts [hunk 12]
```diff
@@ -43,18 +46,34 @@ export class OutgoingWebhookStore extends BaseStore {
...
   @action
   async updateItem(id: OutgoingWebhook['id'], fromOrganization = false) {
-    const response = await this.getById(id, false, fromOrganization);
+    let outgoungWebhook;
+
+    try {
+      outgoungWebhook = await this.getById(id, true, fromOrganization);
+    } catch (error) {
+      if (error.response.data.error_code === 'wrong_team') {
+        outgoungWebhook = {
```
Comment (joeyorlando) [inline grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts:55]: spelling mistake in variable name
```suggestion
    let outgoingWebhook;

    try {
      outgoingWebhook = await this.getById(id, true, fromOrganization);
    } catch (error) {
      if (error.response.data.error_code === 'wrong_team') {
        outgoingWebhook = {
```

File: grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts [hunk 13]
```diff
@@ -43,18 +46,34 @@ export class OutgoingWebhookStore extends BaseStore {
...
+    }
 
-    this.items = {
-      ...this.items,
-      [id]: response,
-    };
+    if (outgoungWebhook) {
+      this.items = {
+        ...this.items,
+        [id]: outgoungWebhook,
```
Comment (joeyorlando) [inline grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts:66]: spelling mistake in variable name
```suggestion
    if (outgoingWebhook) {
      this.items = {
        ...this.items,
        [id]: outgoingWebhook,
```

File: grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts [hunk 14]
```diff
@@ -43,18 +46,34 @@ export class OutgoingWebhookStore extends BaseStore {
...
+
+    try {
+      outgoungWebhook = await this.getById(id, true, fromOrganization);
+    } catch (error) {
+      if (error.response.data.error_code === 'wrong_team') {
+        outgoungWebhook = {
+          id,
+          name: '🔒 Private outgoing webhook',
+          private: true,
+        };
```
Comment (joeyorlando) [inline grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts:59]: can the logic to create this object be be extracted/reused? We're doing something similar [here](https://github.com/grafana/oncall/pull/1528/files#diff-ec407c4a628388542c8d8826e89b579e7f9d99ff085ac263050d3e547ba8c8bfR80-R84) and [here](https://github.com/grafana/oncall/pull/1528/files#diff-5ad2c4830bdc85000eb867815c71d2b6aa5dbcbd4aedc28ad65d9d7f4dd09393R161-R164) (and maybe in a few other spots in the PR). Something like this:

```ts
type MyObject = {
  id: string;  // or maybe number? not sure
  objectName: string;
};

interface PrivateObject extends Pick<MyObject, 'id'> {
  name: string;
  private: boolean;
}

const generatePrivateObject = ({ id, objectName }: MyObject): PrivateObject => ({
  id,
  name `🔒 Private ${objectName}`,
  private: true,
});
```

File: grafana-plugin/src/models/schedule/schedule.ts [hunk 15]
```diff
@@ -153,14 +152,27 @@ export class ScheduleStore extends BaseStore {
...
+      let schedule;
+      try {
+        schedule = await this.getById(id, true, fromOrganization);
+      } catch (error) {
+        if (error.response.data.error_code === 'wrong_team') {
+          schedule = {
+            id,
+            name: '🔒 Private schedule',
+            private: true,
+          };
```
Comment (joeyorlando) [inline grafana-plugin/src/models/schedule/schedule.ts:164]: see comment in `grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts`

File: grafana-plugin/src/models/schedule/schedule.ts [hunk 16]
```diff
@@ -153,14 +152,27 @@ export class ScheduleStore extends BaseStore {
...
-        ...this.items,
-        [item.id]: item,
-      };
-
-      return item;
+      let schedule;
+      try {
+        schedule = await this.getById(id, true, fromOrganization);
+      } catch (error) {
+        if (error.response.data.error_code === 'wrong_team') {
```
Comment (joeyorlando) [inline grafana-plugin/src/models/schedule/schedule.ts:159]: could this `wrong_team` error code be extracted into a constant?

File: grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts [hunk 17]
```diff
@@ -43,18 +46,34 @@ export class OutgoingWebhookStore extends BaseStore {
 
   @action
   async updateItem(id: OutgoingWebhook['id'], fromOrganization = false) {
-    const response = await this.getById(id, false, fromOrganization);
+    let outgoungWebhook;
+
+    try {
+      outgoungWebhook = await this.getById(id, true, fromOrganization);
+    } catch (error) {
+      if (error.response.data.error_code === 'wrong_team') {
```
Comment (joeyorlando) [inline grafana-plugin/src/models/outgoing_webhook/outgoing_webhook.ts:54]: could `wrong_team` be extracted into a constant?

File: grafana-plugin/src/models/escalation_chain/escalation_chain.ts [hunk 18]
```diff
@@ -68,11 +71,38 @@ export class EscalationChainStore extends BaseStore {
   }
 
   @action
-  async updateItems(query = '') {
+  async updateItem(id: EscalationChain['id'], skipErrorHandling = false): Promise<EscalationChain> {
+    let escalationChain;
+    try {
+      escalationChain = await this.getById(id, skipErrorHandling);
+    } catch (error) {
+      if (error.response.data.error_code === 'wrong_team') {
```
Comment (joeyorlando) [inline grafana-plugin/src/models/escalation_chain/escalation_chain.ts:79]: could `wrong_team` be extracted into a constant?

File: grafana-plugin/src/pages/outgoing_webhooks/OutgoingWebhooks.tsx [hunk 19]
```diff
@@ -40,10 +43,6 @@ class OutgoingWebhooks extends React.Component<OutgoingWebhooksProps, OutgoingWe
     errorData: initErrorDataState(),
   };
 
-  async componentDidMount() {
-    this.update().then(this.parseQueryParams);
-  }
```
Comment (joeyorlando) [inline grafana-plugin/src/pages/outgoing_webhooks/OutgoingWebhooks.tsx:45]: what does removing this do?

File: grafana-plugin/src/pages/schedules/Schedules.tsx [hunk 20]
```diff
@@ -150,7 +158,8 @@ class SchedulesPage extends React.Component<SchedulesPageProps, SchedulesPageSta
         <div className={cx('root')}>
           <VerticalGroup>
             <div className={cx('schedules__filters-container')}>
-              <SchedulesFilters value={filters} onChange={this.handleSchedulesFiltersChange} />
+              <RemoteFilters query={query} page="schedules" onChange={this.handleSchedulesFiltersChange} />
+              {/*  <SchedulesFilters value={filters} onChange={this.handleSchedulesFiltersChange} />  */}
```
Comment (joeyorlando) [inline grafana-plugin/src/pages/schedules/Schedules.tsx:162]: can this line be removed?

File: grafana-plugin/src/pages/settings/tabs/TeamsSettings/TeamsSettings.tsx [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+@observer
+class TeamsSettings extends React.Component<AppRootProps, TeamsSettingsPageState> {
+  state: TeamsSettingsPageState = {
+    showEditTeamsModal: false,
+    errorData: initErrorDataState(),
+  };
+  // componentDidMount() {
+  //   const { query } = this.props;
+  //
+  // }
```
Comment (joeyorlando) [inline grafana-plugin/src/pages/settings/tabs/TeamsSettings/TeamsSettings.tsx:24]: can this be removed?

File: engine/apps/api/views/escalation_chain.py [hunk 22]
```diff
@@ -11,10 +12,22 @@
...
 
+class EscalationChainFilter(ByTeamModelFieldFilterMixin, ModelFieldFilterMixin, filters.FilterSet):
+    team = filters.ModelMultipleChoiceFilter(
+        field_name="team",
+        queryset=get_team_queryset,
+        to_field_name="public_primary_key",
+        null_label="noteam",
+        null_value="null",
+        method=ByTeamModelFieldFilterMixin.filter_model_field_with_multiple_values.__name__,
+    )
```
Comment (joeyorlando) [inline engine/apps/api/views/escalation_chain.py:28]: what do you think about deduplicating this filter instance into a new class that subclasses `filters.ModelMultipleChoiceFilter`? This same filter logic is defined in four spots ([here](https://github.com/grafana/oncall/pull/1528/files#diff-17e2f72d153f97776a8ff0dd5edbbc415df4c5d9943f2555ab94221e098530d1R20-R27), [here](https://github.com/grafana/oncall/pull/1528/files#diff-6e0af191419ed818d441082ae583552cb8c1f18e459ff6a593c8cd2b6b7baa6dR63-R70), [here](https://github.com/grafana/oncall/pull/1528/files#diff-4bab66851ffc1e7845a537e83ee5624c99176b2a2c60bf141213919c95c9e6e7R38-R45), and [here](https://github.com/grafana/oncall/pull/1528/files#diff-4582f10ddb8985ec764d973c8de1e598a0b48bff6a2bcd539ba7b093c3dc17c3R98-R105))

File: engine/apps/api/views/alert_group.py [hunk 23]
```diff
@@ -558,7 +578,7 @@ def filters(self, request):
                 "name": "status",
                 "type": "options",
                 "options": [
-                    {"display_name": "new", "value": AlertGroup.NEW},
+                    {"display_name": "firing", "value": AlertGroup.NEW},
```
Comment (joeyorlando) [inline engine/apps/api/views/alert_group.py:581]: could this be done in a separate PR? Are you sure to have updated all other references to "new" to "firing"? If we make this change I think it'd also make sense to refactor `AlertGroup.NEW` -> `AlertGroup.FIRING`

**Review by joeyorlando — COMMENTED** [review]
This is pretty amazing. Seriously, awesome work on this 🎉 🎉 🎉 

A couple broader outstanding thoughts/questions.
- `CHANGELOG.md` - this is a fairly large change, would it make sense to bump to `v1.2.0`?
- public documentation - should we maybe document the implications of this? What adding a resource to a team implies + what "No team" implies?
- integration tests - do you mind adding some basic integration tests for these changes?  some ideas for tests could be:
  - creating the resource types affected by this this PR, with and without a selected team
  - filtering these resource types by one or more teams and making sure only the expected objects are returned
  - updating a user's default team on the new settings page

File: engine/apps/api/views/team.py [hunk 24]
```diff
@@ -17,9 +19,15 @@ class TeamViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
...
-        # dirty hack to render "General" team in team select on the frontend
-        general_team = Team(public_primary_key=None, name="General", email=None, avatar_url=None)
+    def list(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
+        general_team = Team(public_primary_key="null", name="no team", email=None, avatar_url=None)
 
-        return teams + [general_team]
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self.get_serializer([general_team] + list(page), many=True)
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/apps/api/views/team.py:30]: yes, "general" team is available to all users

File: engine/apps/api/views/alert_group.py (hunk 23 above)
Comment (iskhakov) [reply engine/apps/api/views/alert_group.py:581]: I made alert groups page look consistent, also it is called "firing" everywhere in the docs, we can finish that renaming in the separate PR

Screenshot 2023-03-20 at 12 40 46

File: grafana-plugin/src/components/CardButton/CardButton.tsx [hunk 25]
```diff
@@ -21,13 +21,9 @@ const cx = cn.bind(styles);
...
   const { icon, description, title, selected, onClick } = props;
 
-  const handleClick = useCallback(() => {
-    onClick(!selected);
-  }, [selected]);
-
   return (
     <Block
-      onClick={handleClick}
+      onClick={() => onClick(!selected)}
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/components/CardButton/CardButton.tsx:26]: @maskin25 could you take a look please

File: grafana-plugin/src/containers/GSelect/GSelect.tsx [hunk 26]
```diff
@@ -61,9 +62,10 @@ const GSelect = observer((props: GSelectProps) => {
     showWarningIfEmptyValue = false,
     getDescription,
     filterOptions,
-    fromOrganization,
+    // fromOrganization,
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/containers/GSelect/GSelect.tsx:65]: @maskin25 could you take a look

File: grafana-plugin/src/containers/IncidentsFilters/IncidentsFilters.tsx [hunk 27]
```diff
@@ -291,6 +292,21 @@ class IncidentsFilters extends Component<IncidentsFiltersProps, IncidentsFilters
...
 
+      case 'team_select':
+        return (
+          <RemoteSelect
+            autoFocus={autoFocus}
+            className={cx('filter-select')}
+            isMulti
+            fieldToShow="name"
+            valueField="id"
+            href={filter.href.replace('/api/internal/v1', '')}
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/containers/IncidentsFilters/IncidentsFilters.tsx:303]: @maskin25 could you take a look

File: grafana-plugin/src/containers/RemoteFilters/RemoteFilters.tsx
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/containers/RemoteFilters/RemoteFilters.tsx:1]: it renders the whole filter block on the page dynamically based on the `/filters/` endpoint on the backend

File: grafana-plugin/src/containers/TeamName/TeamName.tsx (hunk 6 above)
Comment (iskhakov) [reply grafana-plugin/src/containers/TeamName/TeamName.tsx:29]: I've added it to remove confusion after the update. Otherwise we could silently rename General to "no team"
Screenshot 2023-03-20 at 12 55 02

File: engine/apps/user_management/models/user.py [hunk 28]
```diff
@@ -183,6 +183,12 @@ class Meta:
...
+        if self.role == LegacyAccessControlRole.ADMIN:
+            return self.organization.teams.all()
+        return self.teams.all() | self.organization.teams.filter(is_sharing_resources_to_all=True)
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/apps/user_management/models/user.py:190]: thank you, that's really cool optimisation

File: engine/common/api_helpers/mixins.py [hunk 29]
```diff
@@ -196,6 +196,11 @@ class TeamFilteringMixin:
 
     TEAM_LOOKUP = "team"
 
+    @property
+    def available_teams_lookup_args(self):
+        team_ids = list(self.request.user.available_teams.values_list("id", flat=True))
+        return [Q(**{f"{self.TEAM_LOOKUP}__in": team_ids}) | Q(**{f"{self.TEAM_LOOKUP}__isnull": True})]
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/common/api_helpers/mixins.py:202]: great idea, thank you!

File: grafana-plugin/src/pages/settings/tabs/MainSettings/MainSettings.tsx [hunk 30]
```diff
@@ -65,6 +70,14 @@ class SettingsPage extends React.Component<SettingsPageProps, SettingsPageState>
             </WithPermissionControlTooltip>
           </Field>
         </div>
+        {!isTopNavbar() && (
+          <div style={{ marginBottom: '20px' }}>
+            <Text.Title level={3} className={cx('title')}>
+              Teams and Access Settings
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/pages/settings/tabs/MainSettings/MainSettings.tsx:76]: thank you, done

File: grafana-plugin/src/containers/RemoteFilters/RemoteFilters.tsx [hunk 31]
```diff
@@ -0,0 +1,366 @@
...
+
+    let { filters, values } = parseFilters({ ...query, ...filtersStore.globalValues }, filterOptions);
+
+    if (isEmpty(values)) {
+      let newQuery = defaultFilters || { team: [] };
+      /*  if (filtersStore.values[page]) {
+        newQuery = { ...filtersStore.values[page] };
+      } else {
+        newQuery = defaultFilters || { team: [] };
+      } */
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/containers/RemoteFilters/RemoteFilters.tsx:67]: @maskin25 can you please take a look

File: grafana-plugin/src/containers/TeamsList/TeamsList.tsx (hunk 8 above)
Comment (iskhakov) [reply grafana-plugin/src/containers/TeamsList/TeamsList.tsx:42]: rewrote this part

File: engine/apps/api/views/user.py [hunk 32]
```diff
@@ -84,7 +84,10 @@ def get(self, request):
         return Response(serializer.data)
 
     def put(self, request):
-        serializer = UserSerializer(request.user, data=self.request.data, context={"request": self.request})
+        data = self.request.data
+        if data.get("current_team") == "null":
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/apps/api/views/user.py]: done!

File: engine/apps/user_management/models/team.py [hunk 33]
```diff
@@ -79,3 +79,5 @@ class Team(models.Model):
     name = models.CharField(max_length=300)
     email = models.CharField(max_length=300, null=True, blank=True, default=None)
     avatar_url = models.URLField()
+
+    is_sharing_resources_to_all = models.BooleanField(default=False)
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/apps/user_management/models/team.py:85]: done

File: grafana-plugin/src/containers/TeamsList/TeamsList.tsx (hunk 7 above)
Comment (iskhakov) [reply grafana-plugin/src/containers/TeamsList/TeamsList.tsx:30]: answered above

**Comment by iskhakov** [comment]
@vadimkerr thank you for the thorough review

* When loading a page with the team filter pre-selected, it first loads all resources from org and only then re-renders the page with relevant results, which makes all the pages look a bit glitchy. Is this expected?

Maxim is catching up on that

* Does the user page show all users from org? I'm not sure which users are shown and there's no info on that. Probably treating users as all other resources would make sense? For example, it could be useful to filter users by team and see their notification policies IMO.
It is now showing all the users, you're thinking in the right direction, it is going to be the next goal

* Will the old frontend version work well with the new backend?
Old versions of backend will show all the resources user has access to, without filtering abiity

File: grafana-plugin/src/pages/settings/tabs/MainSettings/MainSettings.tsx (hunk 30 above)
[the remark this answers is not in the document]
Comment (iskhakov) [reply grafana-plugin/src/pages/settings/tabs/MainSettings/MainSettings.tsx:76]: done

File: grafana-plugin/src/pages/schedules/Schedules.tsx (hunk 20 above)
Comment (iskhakov) [reply grafana-plugin/src/pages/schedules/Schedules.tsx]: done

File: grafana-plugin/src/pages/settings/tabs/TeamsSettings/TeamsSettings.tsx (hunk 21 above)
Comment (iskhakov) [reply grafana-plugin/src/pages/settings/tabs/TeamsSettings/TeamsSettings.tsx]: done

File: engine/apps/api/views/schedule.py [hunk 34]
```diff
@@ -219,6 +227,7 @@ def get_object_from_organization(self):
         organization = self.request.auth.organization
         queryset = organization.oncall_schedules.filter(
             public_primary_key=pk,
+            *self.available_teams_lookup_args,
```
Comment (Ferril) [inline engine/apps/api/views/schedule.py:230]: Same case as with custom buttons. This method is used to show selected object in escalation step if this object belongs to other team (case when escalation chain was moved to another team). I don't think that we should filter by available teams here 🤔

File: engine/apps/api/views/team.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/api/views/team.py:30]: This piece of code seems to introduce the "General" team on every page of the response, e.g. on the first page there will be "General" team and some others, on the next page there will be "General" team again and some others, and so on. Could this lead to some unexpected behaviour in the UI when there are multiple "General" teams?

File: grafana-plugin/src/components/CardButton/CardButton.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (teodosii) [reply grafana-plugin/src/components/CardButton/CardButton.tsx:26]: This is a 1 liner, it even adds more complexity if used with `useCallback` compared to inlining it.

See this https://dmitripavlutin.com/react-usecallback/#4-a-bad-use-case

File: grafana-plugin/src/containers/AlertRules/AlertRules.tsx [hunk 35]
```diff
@@ -708,20 +708,19 @@ class AlertRules extends React.Component<AlertRulesProps, AlertRulesState> {
                   {channelFilterIds.length > 1 && <Text keyboard>ELSE</Text>}
                   <Text>route to escalation chain:</Text>
                   <WithPermissionControlTooltip userAction={UserActions.IntegrationsWrite}>
-                    <div onClick={(e) => e.stopPropagation()}>
```
Comment (teodosii) [inline grafana-plugin/src/containers/AlertRules/AlertRules.tsx:711]: Why the removal of `stopPropagation` ? There's no comment indicating why we're doing that but I assume it's fix for a different scenario.

File: engine/apps/api/views/schedule.py (hunk 34 above)
Comment (iskhakov) [reply engine/apps/api/views/schedule.py:230]: we will display this kind of schedules as "🔒private schedule"

File: engine/apps/api/views/custom_button.py [hunk 36]
```diff
@@ -47,9 +62,8 @@ def get_object_from_organization(self):
         # use this method to get the object from the whole organization instead of the current team
         pk = self.kwargs["pk"]
         organization = self.request.auth.organization
-
         try:
-            obj = organization.custom_buttons.get(public_primary_key=pk)
+            obj = organization.custom_buttons.filter(*self.available_teams_lookup_args).get(public_primary_key=pk)
```
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/apps/api/views/custom_button.py:66]: we will display this kind of schedules as "🔒private outgoing webhook"

[the remark this answers is not in the document]
Comment (Ferril) [reply engine/apps/api/views/custom_button.py:66]: got it, good idea 👍

File: engine/apps/api/views/paging.py [hunk 37]
```diff
@@ -20,19 +20,22 @@ class DirectPagingAPIView(APIView):
...
+            data=request.data, context={"organization": organization, "request": request}
+        )
         serializer.is_valid(raise_exception=True)
 
         users = [(user["instance"], user["important"]) for user in serializer.validated_data["users"]]
         schedules = [
             (schedule["instance"], schedule["important"]) for schedule in serializer.validated_data["schedules"]
         ]
 
+        print(serializer.validated_data["team"])
```
Comment (joeyorlando) [inline engine/apps/api/views/paging.py:34]: `print` statement can be removed

File: engine/common/api_helpers/mixins.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (iskhakov) [reply engine/common/api_helpers/mixins.py]: admins now have access to all teams, so I had to modify this suggestion
